File: Python/20. Valid Parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        left = '{[('
        right = {'}':'{',']':'[',')':'('}
        
        for c in s:
            if c in left:
                stack.append(c)
            elif c in right:
                if stack and right[c] == stack[-1]:
                    stack.pop()
                else:
                    return False
        
        return True if len(stack)==0 else False

if __name__ == '__main__':
    s = '{[()]}'
    result = Solution().isValid(s)
    print(result)

File: CPP/20. Valid Parentheses.cpp
#include <iostream>
#include <string>
#include <map>
#include <vector>
#include "printFunctions.hpp"

using namespace std;

class Solution {
public:
	bool isValid(string s) {
		vector<char> stack;
		string left = "{[(";
		map<char, char> right = { {'}','{'},{']','['},{')','('} };

		for (char ch : s) {
			if ( left.find(ch)!=string::npos) {
				stack.push_back(ch);
			}
			else if (right.find(ch)!=right.end()) {
				if (!stack.empty() && right[ch] == stack.back()) {
					stack.pop_back();
				}
				else {
					return false;
				}
			}
		}
		return stack.empty();
	}
};

int main() {
	Solution sol;
	string s = "{[()]}";
	bool result = sol.isValid(s);
	printBoolean(result);
	return 0;
}